Indicators/ImpliedVolatility.cs

```diff
@@ -0,0 +1,226 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using MathNet.Numerics.RootFinding;
+using QuantConnect.Data.Consolidators;
+using QuantConnect.Data.Market;
+using QuantConnect.Logging;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// Implied Volatility indicator that calculate the IV of an option using Black-Scholes Model
+    /// </summary>
+    public class ImpliedVolatility : BarIndicator, IIndicatorWarmUpPeriodProvider
+    {
+        private readonly Symbol _optionSymbol;
+        private readonly Symbol _underlyingSymbol;
+        private BaseDataConsolidator _consolidator;
+        private RateOfChange _roc;
+        private decimal _impliedVolatility;
+        private OptionPricingModelType _optionModel;
+
+        /// <summary>
+        /// Gets the expiration time of the option
+        /// </summary>
+        public DateTime Expiry { get; }
+
+        /// <summary>
+        /// Gets the option right (call/put) of the option
+        /// </summary>
+        public OptionRight Right { get; }
+
+        /// <summary>
+        /// Risk Free Rate
+        /// </summary>
+        public decimal RiskFreeRate { get; set; }
+
+        /// <summary>
+        /// Gets the strike price of the option
+        /// </summary>
+        public decimal Strike { get; }
+
+        /// <summary>
+        /// Gets the option style (European/American) of the option
+        /// </summary>
+        public OptionStyle Style { get; }
+
+        /// <summary>
+        /// Gets the historical volatility of the underlying
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> HistoricalVolatility { get; }
+
+        /// <summary>
+        /// Gets the option price level
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> Price { get; }
+
+        /// <summary>
+        /// Gets the underlying's price level
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> UnderlyingPrice { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the ImpliedVolatility class
+        /// </summary>
+        /// <param name="option">The option to be tracked</param>am>
+        /// <param name="riskFreeRate">The risk free rate</param>
+        /// <param name="period">The lookback period of historical volatility</param>
+        /// <param name="optionModel">The option pricing model used to estimate IV</param>
+        public ImpliedVolatility(Symbol option, decimal riskFreeRate = 0.05m, int period = 252,
```

Comment:
This is kind of arbitrary I think. For example, IB seems to use spot 10y yield rate. Yet, it would be nice to have `IRiskFreeInterestRateModel` as default though.👍